### ISSUE TITLE
Star importing module that imports `__future__.annotations` without defining `__all__` causes TypeAliases used externally to be illegal

### ISSUE THREAD

**Issue by SebbyLaw** [microsoft/pyright#4983]
**Describe the bug**

Star importing a module that star imports `__future__.annotations` causes `TypeAlias`es used externally to be illegal

**To Reproduce**

Consider the following project structure:
```
.
├── README.md
├── foolib
│   ├── core
│   │   ├── __init__.py
│   │   └── impl.py
│   └── types.py
├── main.py
├── main2.py
├── poetry.lock
└── pyproject.toml
```

Type Aliases are defined in `foolib/types.py`, which are used internally by the module and externally by other files/modules that use foolib. The `foolib/impl.py` file uses `from __future__ import annotations`, as well as a TypeAlias from the `foolib/types.py` file.

Two external files, `main.py` and `main2.py` both use the same TypeAlias from the types file as the foolib module. Pyright does not recognize it as a legal `TypeAlias`.

The current solution - which I admit developers should already be doing - is just to not use star import notation, or to define an `__all__` for the modules/files being star imported.

**Expected behavior**

Pyright recognizes the TypeAlias as legal.

**Screenshots or Code**

Following code can be found here: https://github.com/SebbyLaw/min-repro-pyright/tree/typealias-illegal

In `foolib/core/impl.py`:
```python
from __future__ import annotations

from foolib.types import BAR

# everything works fine here
class Foo:
    def __init__(self, bar: BAR) -> None:
        self.bar: BAR = bar

    @staticmethod
    def make_foo(bar: BAR) -> Foo:  # self-referential return type requires future annotations
        return Foo(bar)

# other functions, classes, etc.
```

In `foolib/core/__init__.py`:
```py
from .impl import *
```

In `foolib/types.py`:
```py
from typing import Literal, TypeAlias

BAR: TypeAlias = Literal["bar", "baz", "barbaz"]
```

In `main.py`:
```py
from foolib.core import *
from foolib.types import BAR

class Bar:
    def __init__(self, bar: BAR):
        self.foo = Foo(bar)  # okay
        self.bar: BAR = bar  # pyright error on this line

def make_bar(bar: BAR) -> Bar:
    return Bar(bar)  # okay

another_bar: BAR = "bar"  # okay
```

More errors in `main2.py`, with the only difference between `main.py` being the import of `__future__.annotations`:
```py
from __future__ import annotations

from foolib.core import *
from foolib.types import BAR

class Bar:
    def __init__(self, bar: BAR):
        self.foo = Foo(bar)  # okay
        self.bar: BAR = bar  # pyright error on this line

def make_bar(bar: BAR) -> Bar:
    return Bar(bar)  # okay

another_bar: BAR = "bar"  # okay
```

We can actually simplify this example a bit, and get even more errors: https://github.com/SebbyLaw/min-repro-pyright/tree/typealias-illegal-simple
```
.
├── README.md
├── foolib.py
├── footypes.py
├── main.py
├── main2.py
├── poetry.lock
└── pyproject.toml
```

Where `footypes.py` is the original `foolib/types.py` file, and `foolib.py` is the original `foolib/impl.py` file:
```py
from typing import Literal, TypeAlias

BAR: TypeAlias = Literal["bar", "baz", "barbaz"]
```
```py
from __future__ import annotations

from footypes import BAR

# everything works fine here
class Foo:
    def __init__(self, bar: BAR) -> None:
        self.bar: BAR = bar

    @staticmethod
    def make_foo(bar: BAR) -> Foo:  # self-referential return type requires future annotations
        return Foo(bar)

# other functions, classes, etc.
```

Output from running `pyright .`:
```
No configuration file found.
pyproject.toml file found at /home/sebnlaw/min-repro-pyright.
Loading pyproject.toml file at /home/sebnlaw/min-repro-pyright/pyproject.toml
Assuming Python version 3.10
Assuming Python platform Linux
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
Searching for source files
Found 5 source files
pyright 1.1.304
/home/sebnlaw/min-repro-pyright/main.py
  /home/sebnlaw/min-repro-pyright/main.py:8:19 - error: Illegal type annotation: variable not allowed unless it is a type alias (reportGeneralTypeIssues)
/home/sebnlaw/min-repro-pyright/main2.py
  /home/sebnlaw/min-repro-pyright/main2.py:8:29 - error: Illegal type annotation: variable not allowed unless it is a type alias (reportGeneralTypeIssues)
  /home/sebnlaw/min-repro-pyright/main2.py:10:19 - error: Illegal type annotation: variable not allowed unless it is a type alias (reportGeneralTypeIssues)
  /home/sebnlaw/min-repro-pyright/main2.py:13:19 - error: Illegal type annotation: variable not allowed unless it is a type alias (reportGeneralTypeIssues)
  /home/sebnlaw/min-repro-pyright/main2.py:17:14 - error: Illegal type annotation: variable not allowed unless it is a type alias (reportGeneralTypeIssues)
5 errors, 0 warnings, 0 informations
```

Screenshot of `main.py` in the first foolib module example:
![Code_L5l3SlZZZp](https://user-images.githubusercontent.com/44045823/233521974-d4de2214-b7da-4a7a-a662-b9737da46ab0.png)
And `main2.py`:
![Code_aWcSGnsD2p](https://user-images.githubusercontent.com/44045823/233522032-a855a3e2-85b7-4447-82a0-1b275855222e.png)

**VS Code extension or command-line**

```bash
$ pyright --version
pyright 1.1.304
```

VSCode extension info:
```
-   Language Server version: 2023.4.30
-   OS and version: linux x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: basic
```

**Comment by erictraut** [comment]
Thanks for the bug report and the clear repro steps.

This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.305, which I just published. It will also be included in a future release of pylance.
